proposals/redirects.md

```diff
@@ -0,0 +1,156 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 2023-05-22
+- Reference Issues: https://github.com/withastro/roadmap/issues/466
+- Implementation PR: https://github.com/withastro/astro/pull/7067/
+
+# Summary
+
+Add a `redirects` config option to the Astro config which allows you to define redirects in a central place. Allow integrations to read this information and apply it to it's own redirects configuration for deployment.
+
+# Example
+
+New config option used like so:
+
+```js
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+  redirects: {
+    '/other': '/place'
+  }
+});
+```
+
+You can also specify the status code by using an object notation:
+
+```js
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+  redirects: {
+    '/other': {
+      status: 302,
+      destination: '/place'
+    }
+  }
+});
+```
+
+# Background & Motivation
+
+This was original proposed as a stage 1 discussion [here](https://github.com/withastro/roadmap/discussions/319) and was one of the top voted proposals.
+
+As websites age there are times where routes are rearchitectured. In order preserve existing links to content on the web it is common to set up redirects in your web server.
+
+Redirect configuration varies depending on what web server or host you use. Some times you might even want to change hosts, in which case your redirects need to be converted over to a new format.
+
+Having a redirects configuration within Astro itself allows a single place to define redirects that works everywhere.
+# Goals
+
+- Works in development mode.
+- Writes out `<meta http-equiv="refresh">` tags in static builds.
+- Gives proper hooks to integrations so that they can write to their own configuration.
+
+# Non-Goals
+
+- Dynamic behavior that goes beyond the capabilities of the file-based routing system. So nothing based on the properties of the request, user session, etc. Regular routes still should be used for this scenario.
+- Redirects to pages that do not exist in the Astro project.
+- External redirects.
+
+# Detailed Design
+
+This will be implemented as a feature of the internal routing. The `RouteData` type will be extended to include:
+
+```js
+interface RedirectConfig = string | {
+  status: 300 | 301 | 302 | 303 | 304 | 307 | 308;
```

Comment:
Looks like Cloudflare does this too: allows `200` for [“proxying”](https://developers.cloudflare.com/pages/platform/redirects/#proxying) and `404` for [“rewrites”](https://developers.cloudflare.com/pages/platform/redirects/#create-redirects). Pretty similar to Netlify (I forgot they support 404 too).

Reply:
On Vercel redirects and rewrites are handled separately and looks like Netlify/Cloudflare’s `200` redirect is what Vercel calls a “rewrite”: https://vercel.com/docs/concepts/edge-network/rewrites

Reply:
Not really familiar with Deno Deploy and don’t see anything obvious in their docs. I guess there it’s closer to our Node adapter? So up to us how exactly we handle stuff.

Looks like the static site platforms use a “200 redirect” to mean a server-side rewrite where the destination content gets served at the requested URL instead of notifying about a redirect. Probably useful for SPAs in particular.